### ISSUE TITLE
Atlas Texture region ignored when using CanvasTexture and Sprite2D

### ISSUE THREAD

**Issue by RobTF** [godotengine/godot#67936]
### Godot version

4.0.beta3.mono

### System information

Windows 11, Visual Studio 2022

### Issue description

I've got a sprite sheet, which has been split into two `AtlasTexture` instances. If I create a `Sprite2D` and attach a `CanvasTexture` with the intention of adding a diffuse texture with normal map I notice that if I select one of my `AtlasTexture` instances the atlas region is ignored and the sprite always shows the texture from the sprite sheet origin.

This issue is the same for the normal and diffuse textures, I can't vouch for other such as specular as I've never tried using them.

See screenshot below which shows the issue;
![image](https://user-images.githubusercontent.com/4292237/198259490-1483e759-f2bb-4a3b-8929-30e5cc8f56d9.png)

### Steps to reproduce

- Add sprite sheet to project.
- Add `AtlasTexture` resources for each frame.
- Add a `Sprite2D`
- Use a `CanvasTexture` for the new sprite.
- For the diffuse texture, select any `AtlasTexture` frame other than the one aligned with the sprite sheet origin.
- Note that the sprite displays the wrong image.

### Minimal reproduction project

[TextureAtlasBug.zip](https://github.com/godotengine/godot/files/9878441/TextureAtlasBug.zip)

**Comment by MrBlockers** [comment]
Does this occur on both the Vulkan renderer and the OpenGL3 / gl_compatibility renderer?

**Comment by RobTF** [comment]
@MrBlockers I _think_ so, but I'm uncertain how to test it I'm afraid. I tried setting the rendering method to all three options, `forward_plus`, `mobile` and `gl_compatability`. All seem to be the same, but notably none say `vulkan` although one or more of these may use that as some sort of backend perhaps?

**Comment by MrBlockers** [comment]
I think forward_plus always uses vulkan, it's laid out a little strangely in the options imo. It sounds like it's not a renderer specific issue at least!

I did a little investigation and I think the underlying issue is because AtlasTexture's cropping behavior doesn't seem to actually crop the texture in memory. It's sort of just a wrapper around another Texture2D to create cropped draw commands.

**Comment by Mickeon** [comment]
> I did a little investigation and I think the underlying issue is because AtlasTexture's cropping behavior doesn't seem to actually crop the texture in memory. It's sort of just a wrapper around another Texture2D to create cropped draw commands.

Yeah, no new texture is created and it's all fetched from the atlas. It's a clever and very useful memory optimisation but it'll frequently have some sort of oddity like this...

**Comment by Calinou** [comment]
AtlasTexture needs explicit support in every node/resource that accepts it to work.

**Comment by adamscoble** [comment]
Hi all! Just came across this issue myself. 

Would creating an alternate shader for the CanvasTexture be able to solve this, so I can use AtlasTextures with normal maps? 

I could store the Rect I was using for the AtlasTexture and apply it to Sprite2D.region_rect, but I'm not sure if that will work for all of the CanvasTexture's textures.

**Comment by LeoArr** [comment]
Hello!
Bumped into this bug as well. The AtlasTexture has the right width and height but the x and y does not move the cropped area.

**Comment by LeoArr** [comment]
Update:
I'm not sure if this an actual workaround but if I enable the region for the Sprite2D and set the Rect to the same values as the AtlasTexture's rect the sprite shows up. I have not tried actually applying this and also use normal maps yet but I think then the normal map would have to have the relevant image reagion in the same place as the original image.

**Comment by CoffeeCatRailway** [comment]
> Update: I'm not sure if this an actual workaround but if I enable the region for the Sprite2D and set the Rect to the same values as the AtlasTexture's rect the sprite shows up. I have not tried actually applying this and also use normal maps yet but I think then the normal map would have to have the relevant image reagion in the same place as the original image.

Thank you! That works, but I found if you do that you don't need to use an AtlasTexture at all as the Sprite2D region will do that anyway.

**Comment by pleyland** [comment]
This line fixed everything for me when adding textures to a CanvasTexture programmatically from an atlas texture:

`sprite_2D.region_rect = sprite_2D.texture.diffuse_texture.region`

Hope it helps everyone else. It basically just grabs the region from the atlas texture and applies it to the Sprite 2D. Not sure if this is really a bug or the intended method
